src/test/util/buffer.rs

```diff
@@ -0,0 +1,464 @@
+    pub(crate) fn get_successful_command_execution(
+        &mut self,
+        command_name: &str,
+    ) -> (CommandStartedEvent, CommandSucceededEvent) {
+        let mut started = None;
+        let mut succeeded = None;
+        self.retain(|ev| match (ev, &started, &succeeded) {
+            (Event::Command(cev), None, None) => {
+                if cev.command_name() == command_name {
+                    started = Some(
+                        cev.as_command_started()
+                            .unwrap_or_else(|| {
+                                panic!("first event not a command started event {:?}", cev)
+                            })
+                            .clone(),
+                    );
+                }
+                false
+            }
+            (Event::Command(cev), Some(started), None) => {
+                if cev.request_id() == started.request_id {
+                    succeeded = Some(
+                        cev.as_command_succeeded()
+                            .expect("second event not a command succeeded event")
+                            .clone(),
+                    );
+                }
+                false
+            }
+            _ => true,
+        });
+        match (started, succeeded) {
+            (Some(started), Some(succeeded)) => (started, succeeded),
+            _ => panic!("could not find event for {} command", command_name),
+        }
+    }
+
+    pub(crate) fn count_pool_cleared_events(&self) -> usize {
+        let mut out = 0;
+        for event in self.all().iter() {
+            if matches!(event, Event::Cmap(CmapEvent::PoolCleared(_))) {
+                out += 1;
+            }
+        }
+        out
+    }
+}
+
+pub(crate) struct EventSubscriber<'a, T> {
+    buffer: &'a EventBuffer<T>,
+    index: usize,
+    generation: Generation,
+}
+
+impl<'a, T: Clone> EventSubscriber<'a, T> {
+    async fn next(&mut self, timeout: Duration) -> Option<T> {
+        crate::runtime::timeout(timeout, async move {
+            loop {
+                let notified = self.buffer.inner.event_received.notified();
+                if let Some(next) = self.try_next() {
+                    return Some(next);
+                }
+                notified.await;
+            }
+        })
+        .await
+        .unwrap_or(None)
+    }
+
+    fn try_next(&mut self) -> Option<T> {
+        let events = self.buffer.inner.events.lock().unwrap();
+        if events.generation != self.generation {
+            panic!("EventBuffer cleared during EventStream iteration");
+        }
+        if events.data.len() > self.index {
+            let event = events.data[self.index].0.clone();
+            self.index += 1;
+            return Some(event);
+        }
+        None
+    }
+
+    /// Consume and pass events to the provided closure until it returns Some or the timeout is hit.
+    pub(crate) async fn filter_map_event<F, R>(
```

Comment:
This and the following methods are preserved from the previous `EventSubscriber` implementation; to avoid introducing bugs I left the implementation as is, but it's likely there are much more efficient ways to implement most of these given the underlying cursor representation.